lib/plugin-manager.js

```diff
@@ -1,26 +1,38 @@
 import { extname, basename } from 'path';
 import { flattenAndCompact } from './utils/array';
 
+function updatePluginsForKey(lookup, plugin, key) {
+  const pluginsForKey = lookup.get(key) || [];
+
+  if (pluginsForKey.includes(plugin)) {
+    return;
+  }
+
+  pluginsForKey.push(plugin);
+
+  if (lookup.has(key)) {
+    return;
+  }
+
+  lookup.set(key, pluginsForKey);
+}
+
 function buildPluginCache(plugins) {
   const lookup = new Map();
 
   plugins.forEach((PluginClass) => {
     const plugin = new PluginClass();
 
-    plugin.getPattern().forEach((pattern) => {
-      const pluginsForPattern = lookup.get(pattern) || [];
-
-      if (pluginsForPattern.includes(plugin)) {
-        return;
-      }
-
-      pluginsForPattern.push(plugin);
+    plugin.getPattern().pathSubstrings.forEach((pattern) => {
```

Comment:
This is the right place where we can distinguish between the plugin interface (an array) and the pattern-preset object interface.